From: SJ Park <sj@kernel.org>
Date: Tue, 4 Aug 2026 07:37:00 -0700
Subject: [PATCH] mm/damon/core: document damon_call()/damon_start() race hang
 issue

When damon_call() is called in parallel to damon_start() for same ctx,
damon_call() could show ctx->damon_calls_obsolte set while ctx->kdamond
is unset.  If ctx->kdamond is set before damon_call() moved to the
cancel stage and wait kdamond stop, damon_call() can indefinitely hang.
There is no such race.  Comment for caution, though.

Signed-off-by: SJ Park <sj@kernel.org>
---
 mm/damon/core.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index a283e9ae6ff9d..abb30819d48c0 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -2383,6 +2383,9 @@ int damon_kdamond_pid(struct damon_ctx *ctx)
  *
  * When this function is failed, the @ctx is guaranteed to be stopped.
  *
+ * This function should not be called in parallel to damon_start() for the
+ * @ctx.  In the case, this function could indefinitely hang.
+ *
  * Return: 0 on success, negative error code otherwise.
  */
 int damon_call(struct damon_ctx *ctx, struct damon_call_control *control)
-- 
2.47.3

